### ISSUE TITLE
Obtained baseline accuracy for ResNet50v2 is different than officially reported

### ISSUE THREAD

**Issue by gcunhase** [keras-team/keras#15822]
Reposting urgent issue from the `keras-applications` repo from 12 days ago (no activity so far, no one has been assigned to it yet): https://github.com/keras-team/keras-applications/issues/207

Also posted the [issue on TF repo](https://github.com/tensorflow/models/issues/10420), but they said the Keras team is responsible for it.

Please check this bug as it is currently a huge bottleneck.

### Summary
Obtained baseline accuracy for ResNet50v2 is different than officially reported:

| Model | Top-1 (%) |
| -------- | ------------- |
| [Baseline (official)](https://github.com/keras-team/keras-applications) | 75.96 |
| Baseline (obtained) | 66.85 |

ResNet50v1 works fine, so I don't think the issue is with data-loading or with the eval code. My current guess is that there might be some disconnect between the loaded checkpoints and the input pre-processing function in ResNetv2 (`keras.applications.resnet_v2.preprocess_input`).

Any advice is appreciated.

### Environment
- Python version: 3.8
- Keras backend with version: TF 2.7

### Logs or source codes for reproduction
1. Download ImageNet `tfrecord` validation dataset.
2. Run evaluation script:
```python
import os
import tensorflow as tf

DATA_DIR = "/media/Data/ImageNet/train-val-tfrecord"
_DEFAULT_IMAGE_SIZE = 224
_NUM_CHANNELS = 3
_RESIZE_MIN = 256


def load_data(data_dir='./data/imagenet', batch_size=8, num_val_files=128):
    # 1. Load ImageNet2012 train dataset - needs to manually download the full ImageNet2012 dataset first.
    assert os.path.exists(data_dir)

    # 2. Make train/validation datasets
    filenames = get_filenames(data_dir, num_val_files=num_val_files)
    dataset = tf.data.TFRecordDataset(filenames)
    dataset = dataset.map(map_func=preprocess_image_record_tf_v2)
    dataset = dataset.batch(batch_size, drop_remainder=True)

    return dataset


def preprocess_image_record_tf_v2(record):
    """
        This function is called by "dataset.map()" in the "map_func" argument.
            That function doesn't allow for multiple input arguments, only "record".
    """
    imgdata, label, bbox, text = _deserialize_image_record(record)
    # Subtract one so that labels are in [0, 1000)
    label -= 1

    try:
        image = tf.image.decode_jpeg(imgdata, channels=_NUM_CHANNELS, fancy_upscaling=False, dct_method='INTEGER_FAST')
    except:
        image = tf.image.decode_image(imgdata, channels=_NUM_CHANNELS)

    # V2
    image = tf.cast(image, tf.float32)
    image = tf.image.resize(image, (_DEFAULT_IMAGE_SIZE, _DEFAULT_IMAGE_SIZE))
    image = tf.keras.applications.resnet_v2.preprocess_input(image)

    return image, label


def get_filenames(data_dir, num_val_files=128):
    """Return filenames for dataset."""
    return [
        os.path.join(data_dir, 'validation-{:05d}-of-{:05d}'.format(i, num_val_files))
        for i in range(num_val_files)
    ]


def _deserialize_image_record(record):
    feature_map = {
        'image/encoded': tf.io.FixedLenFeature([], tf.string, ''),
        'image/class/label': tf.io.FixedLenFeature([], tf.int64, -1),
        'image/class/text': tf.io.FixedLenFeature([], tf.string, ''),
        'image/object/bbox/xmin': tf.io.VarLenFeature(dtype=tf.float32),
        'image/object/bbox/ymin': tf.io.VarLenFeature(dtype=tf.float32),
        'image/object/bbox/xmax': tf.io.VarLenFeature(dtype=tf.float32),
        'image/object/bbox/ymax': tf.io.VarLenFeature(dtype=tf.float32)
    }
    with tf.name_scope('deserialize_image_record'):
        obj = tf.io.parse_single_example(record, feature_map)
        imgdata = obj['image/encoded']
        label = tf.cast(obj['image/class/label'], tf.int32)
        bbox = tf.stack([obj['image/object/bbox/%s' % x].values for x in ['ymin', 'xmin', 'ymax', 'xmax']])
        bbox = tf.transpose(tf.expand_dims(bbox, 0), [0, 2, 1])
        text = obj['image/class/text']

        return imgdata, label, bbox, text


def main():
    # Load tfrecord data
    val_batches = load_data(data_dir=DATA_DIR, batch_size=128)

    # Instantiate Baseline model
    model = tf.keras.applications.ResNet50V2(
            include_top=True, weights='imagenet', input_tensor=None,
            input_shape=None, pooling=None, classes=1000,
            classifier_activation='softmax'
        )

    model.compile(optimizer="sgd", loss=tf.keras.losses.SparseCategoricalCrossentropy(), metrics=["accuracy"])
    _, baseline_model_accuracy = model.evaluate(val_batches)
    print("Baseline val accuracy:", baseline_model_accuracy)


if __name__ == '__main__':
    main()
```

**Comment by gcunhase** [comment]
Is there any update for this bug?

**Comment by innat** [comment]
@gcunhase have you checked on some small-scale data set like mnist/cifar-10/100? Have you faced the same abnormality?

**Comment by qlzh727** [comment]
@gcunhase, I was able to reproduce 0.7630 on imagenet2012 validation set with resnet50_v2 on my local colab, so I am not sure what's the issue you are facing (since I can access the data you are using). 

From the code snippet you provide, I think the only place that I can think of for error is the preprocessing/resize of the image.

My code to produce the output:
# Note that I have to use batch size == 1 below since each of the image in the dataset has different size.
```
import tensorflow as tf
import tensorflow_datasets as tfds
from tensorflow.keras.applications import resnet_v2

def eval_preprocess(features):
  """Process a single example for evaluation."""
  image = features["image"]
  label = features["label"]
  image = tf.cast(image, tf.float32)
  image = resnet_v2.preprocess_input(image)
  return image, label


model = tf.keras.applications.resnet_v2.ResNet50V2(
    include_top=True,
    weights='imagenet',
    classes=1000
    )

model.compile(metrics=['accuracy'])


dataset_builder = tfds.builder("imagenet_2012")
eval_ds = dataset_builder.as_dataset(split='validation')
eval_ds = eval_ds.map(eval_preprocess).batch(1)

_, acc = model.evaluate(eval_ds, verbose=1)
```

**Comment by gcunhase** [comment]
Thank you for your reply @qlzh727 !

We're hoping to use it for fine-tuning, so could you please also try with batch higher than 1? That's the reason we were using resize.

**Comment by qlzh727** [comment]
Did a few more experiment with resizing of the image, on resnet v1 and v2:

V1 with batch size 128
image = tf.image.resize(image, [224, 224]) -> 0.7057
image = tf.image. resize_with_pad(image, 224, 224) -> 0.6771
image (resize with same ratio and center crop to 224, 224) -> 0.7432

V2 with batch size 128
image = tf.image.resize(image, [224, 224]) -> 0.6686
image = tf.image. resize_with_pad(image, 224, 224) -> 0.6226
image (resize with same ratio and center crop to 224, 224) -> 0.6928

There is indeed a big dip in the accuracy, which I don't understand why at the moment. Let me take a closer look.

```
def resize_and_center_crop(image: tf.Tensor, size: int) -> tf.Tensor:
  h, w = tf.shape(image)[0], tf.shape(image)[1]

  # Figure out the necessary h/w.
  ratio = (tf.cast(size, tf.float32) / tf.cast(tf.minimum(h, w), tf.float32))
  h = tf.cast(tf.round(tf.cast(h, tf.float32) * ratio), tf.int32)
  w = tf.cast(tf.round(tf.cast(w, tf.float32) * ratio), tf.int32)
  image = tf.image.resize(image, [h, w])
  top = (h - size) // 2
  left = (w - size) // 2
  image = tf.image.crop_to_bounding_box(image, top, left, size, size)
  return image
```

**Comment by qlzh727** [comment]
Found more details after digging some history.

1. From [keras-team/keras-application](https://github.com/keras-team/keras-applications), I think the original weights we got is from the tensorflow/model slim directory. See [this link](https://github.com/tensorflow/models/blob/master/research/slim/nets/resnet_v2.py)
2. On the slim readme.md, there is a fine print for ResnetV2. The key info here is that the model weights are produced with input shape [299, 299, 3]
```
^ ResNet V2 models use Inception pre-processing and input image size of 299 
(use --preprocessing_name inception --eval_image_size 299 when using eval_image_classifier.py).
 Performance numbers for ResNet V2 models are reported on the ImageNet validation set.
```
3. The same shape information was also showing up on keras-application readme.
```
              Input	Top-1	Top-5	Size	Stem	References
ResNet50V2	299	75.960	93.034	25.6M	23.6M	[paper] [tf-models] [torch]
```

4. When we port the keras-application model to keras-team/keras under keras.application.renset_v2, we force the input shape to be same as the v1 implementation which is [224, 224, 3], but still use the weights which expecting [299, 299, 3]. 

This is main reason that the current ResNetV2 (50/101/152) was not producing same performance number stated in the original doc.

To verify my find, I update the resnet code with following patch to force the mode input shape to be [299, 299, 3]. With that the model is able to produce reasonable result.

V2 with batch size 128
image = tf.image.resize(image, [299, 299]) -> 0.7415
image = tf.image. resize_with_pad(image, 299, 299) ->  0.7107
image (resize with same ratio and center crop to 299, 299) -> 0.7530

```
from keras.applications import imagenet_utils

def patch_with_fixed_shape(*args, **kwargs):
  return (299, 299, 3)
imagenet_utils.obtain_input_shape = patch_with_fixed_shape
```

**Comment by gcunhase** [comment]
@qlzh727 Thank you so much for investigating this issue!!

I am now able to get 74.15% accuracy with `resize` and 75.36% with `cental_crop` if I load the model as follows:

```python
model = tf.keras.applications.resnet_v2.ResNet50V2(
    include_top=True,
    weights='imagenet',
    input_tensor=tf.keras.layers.Input(shape=(299, 299, 3)),
    classes=1000
    )
```

Can you please confirm if this is the correct way to load the model with a different input shape?

**Comment by qlzh727** [comment]
Yea, you can do that. We are actually planning to fix the weights, so that user don't have to change their client side code (eg reshape their input pipeline), so you solution should work for now, but might need to change back when we update weights. 

We will post any fix in this issue.

**Comment by gcunhase** [comment]
Thank you for your help @qlzh727 ! Changing the input_tensor to 299x299x3 seems to solve the issue.
